### PR TITLE
[MOD-14955] Add DistanceCalculatorWithNorm

### PR DESCRIPTION
**Pull request by dor-forer**
**Describe the changes in the pull request**

This PR is stacked on [#1000](https://github.com/RedisAI/VectorSimilarity/pull/1000). Review only the incremental diff against `dor-forer-MOD-14952-support-quant-normalization`.

Cherry-picks the two commits from [ARM-software/VectorSimilarity-for-Arm#3](https://github.com/ARM-software/VectorSimilarity-for-Arm/pull/3): [`493ee7a8`](https://github.com/ARM-software/VectorSimilarity-for-Arm/commit/493ee7a8596c19c10ad2b467364fbfefd61732f6) and [`c27401d6`](https://github.com/ARM-software/VectorSimilarity-for-Arm/commit/c27401d6d5a29e2f98d8c80fbe77f8ce9e57156e).

This PR adds `DistanceCalculatorWithNorm`, which invokes the selected SQ8 kernel and applies the mean-normalization correction terms needed to recover the distance between the original vectors. The follow-up commit makes metadata reads safe for unaligned storage.

The calculator integrates with the construction-time dispatch introduced in #999:

- HNSW caches a non-virtual callback and a minimal immutable context containing only the stored/query kernels and `mean_sum_squares`.
- The candidate loop does not call `IndexCalculatorInterface::calcDistance` or `calcDistanceForQuery`, so it does not restore the per-distance vtable lookup removed by #937.
- Existing calculators retain the stateless raw-kernel path. Stored-to-stored normalized L2 also uses that path because its mean terms cancel.

Tests cover FP32/FP16, L2/IP, symmetric/asymmetric, zero-mean, randomized-vector, and stateless/stateful dispatch behavior.

**Which issues this PR fixes**

1. MOD-14955

**Main objects this PR modified**

1. `DistanceCalculatorWithNorm`
2. Normalized-distance component unit tests

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

**Validation**

- `make format`
- `make check-format`
- Release build of `test_components` and `test_hnsw`
- `test_components`: 40/40 passed
- `test_hnsw` excluding the two external serialization fixtures: 169/169 passed
- `git diff --check dor-forer-MOD-14952-support-quant-normalization...HEAD`
- Focused stateless-dispatch microbenchmark: 36 alternating-order rounds, 20M calls per path per round; cached raw kernel averaged 2.75 ns/call and `DistanceDispatch` averaged 2.69 ns/call (no measurable regression within run-to-run noise)

---

> [!NOTE]
> **Medium Risk**
> Changes core VecSim distance math and blob layouts for normalized SQ8 indices; mistakes would skew ANN search, but behavior is heavily covered by new unit tests including numerical edge cases.
> 
> **Overview**
> Adds **`DistanceCalculatorWithNorm`**, which runs the existing SQ8 distance kernels on mean-centered quantized storage and applies **IP-only** correction terms (`x_mean_ip`, `y_mean_ip`, `mean_sum_squares`) so reported distances match the **original** (unshifted) vectors. **Normalized L2** skips runtime correction: queries are **centered at preprocess time** (`y - mean`), so stored/query kernels already compute the true L2 distance; **L2** paths expose **stateless** `DistanceDispatch` (raw kernels), while **IP** uses **stateful** dispatch for the correction callbacks.
> 
> **`QuantPreprocessor` (WithNorm)** is aligned with that split: L2 query blobs store centered values and metadata sums over centered data; IP keeps raw query values and records **`y_mean_ip`** from the original query. **`sq8` metadata layout** only adds the extra **`mean_ip`** slot when **WithNorm + IP** (not for normalized L2).
> 
> Unit tests cover FP32/FP16, L2/IP, symmetric/asymmetric paths, zero mean, large-offset L2 stability, random vectors, odd dimensions (unaligned metadata reads via `load_unaligned`), and dispatch stateless vs stateful behavior.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 385741f98f53fc00f61110f2c8c759bf03c0eedf. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/VecSim/vec_sim_index.h [hunk 1]
```diff
@@ -163,28 +163,28 @@ struct VecSimIndexAbstract : public VecSimIndexInterface {
      *
      * Uses the cached dist func to avoid the indexCalculator vtable on the hot path.
      *
-     * @note Precondition: @c cachedDistFunc must be non-null. Subclasses that construct
-     *       this index with a null @c indexCalculator (e.g. SVS, which uses its own
-     *       internal distance kernels) must not call this method.
-     *
      * @return the distance between the vectors.
      */
     DistType calcDistance(const void *vector_data1, const void *vector_data2) const {
```
Comment (ofiryanai) [inline src/VecSim/vec_sim_index.h:175]: This method calls `indexCalculator->calcDistance`, restoring the per-distance `IndexCalculatorInterface` vtable fallback removed by PR #937. `DistanceCalculatorWithNorm` always returns nullptr for the query kernel, so every visited candidate now takes a virtual call followed by the raw kernel call; existing calculators also pay the new branch. Cache a non-virtual state-aware callback or specialize the normalized dispatch outside the candidate loop, and benchmark against the #937 path before merging

**Review by ofiryanai — CHANGES_REQUESTED** [review]

File: src/VecSim/spaces/computer/calculator.h [hunk 2]
```diff
@@ -120,3 +122,121 @@ class DistanceCalculatorCommon
...
+            // IP(x, y) = IP(x', y) + y_mean_ip
+            // => 1 - IP(x, y) = base - y_mean_ip
+            return base - y_mean_ip;
+        } else { // L2
+            // base = ||x' - y||². We want ||x - y||².
+            // ||x - y||² = ||x' - y||² + 2*(x_mean_ip - y_mean_ip) - mean_sum_squares
+            float x_mean_ip =
+                load_unaligned<float>(static_cast<const uint8_t *>(candidate) + dim +
+                                      sq8::template mean_ip_index<Metric>() * sizeof(float));
+            return base + 2.0f * (x_mean_ip - y_mean_ip) - context->mean_sum_squares;
```
Comment (ofiryanai) [inline src/VecSim/spaces/computer/calculator.h]: This L2 correction subtracts several `O(||mean||²)` FP32 values after the base kernel has already rounded its result, causing catastrophic cancellation for nearby vectors with a nontrivial mean. For dim=128, mean[i]=1000, x[i]=1001, and y[i]=1001.1, the current FP32 evaluation produces approximately 40 instead of 1.28 and may even become negative. Avoid forming `||x′−y||²` and correcting it afterward; preprocess/use the centered query y−mean for normalized L2, or fuse the calculation so all terms remain centered. Add a large-offset/small-distance regression test.

File: src/VecSim/spaces/computer/calculator.h [hunk 3]
```diff
@@ -120,3 +122,121 @@ class DistanceCalculatorCommon
...
+        spaces::dist_func_t<DistType> query_func;
+        float mean_sum_squares;
+    };
+
+    const WithNormDistanceContext context_;
+
+    static DistType calcStoredWithContext(const void *opaque_context, const void *v1,
+                                          const void *v2, size_t dim) {
+        const auto *context = static_cast<const WithNormDistanceContext *>(opaque_context);
+        DistType base = context->stored_func(v1, v2, dim);
```
Comment (ofiryanai) [inline src/VecSim/spaces/computer/calculator.h:171]: `load_unaligned` protects the new correction metadata, but `stored_func`/`query_func` still read existing SQ8 float metadata through `reinterpret_cast<const float *>(blob + dim)`. That address is unaligned whenever `dim % 4 != 0`, so normalized distances still have alignment UB despite this PR’s unaligned-load fix. Convert the underlying SQ8 kernel metadata reads to `load_unaligned` (or guarantee an aligned layout) and add an odd-dimension sanitizer test; all current calculator tests use dim=8 or 16.

**Review by ofiryanai — CHANGES_REQUESTED** [review]

File: src/VecSim/spaces/computer/calculator.h (hunk 2 above)
Comment (dor-forer) [reply src/VecSim/spaces/computer/calculator.h]: Good catch — this is not fixed in the current head. The contextual dispatch change addressed the per-distance virtual-call regression, but the normalized L2 query path still applies this cancellation-prone correction. I will change normalized L2 query preprocessing to use the centered query (`y - mean`) and add the large-offset/small-distance regression test before resolving this thread.

File: src/VecSim/spaces/computer/calculator.h (hunk 3 above)
Comment (dor-forer) [reply src/VecSim/spaces/computer/calculator.h:171]: This underlying SQ8 metadata-alignment issue is covered by #1001 (MOD-15303), which converts the scalar and SIMD SQ8-FP32/SQ8-SQ8 metadata reads to `load_unaligned<float>` and adds odd-dimension sanitizer regression coverage. We will merge that fix separately and rebase this PR onto it rather than duplicating the broad kernel changes here. I will also add a small odd-dimension `DistanceCalculatorWithNorm` integration test in this PR after rebasing.

File: src/VecSim/spaces/computer/preprocessors.h [hunk 4]
```diff
@@ -370,10 +384,10 @@ class QuantPreprocessor : public PreprocessorInterface {
             }
 
             if constexpr (WithNorm) {
-                m0 += mean[i + 0] * y0;
-                m1 += mean[i + 1] * y1;
-                m2 += mean[i + 2] * y2;
-                m3 += mean[i + 3] * y3;
+                m0 += mean[i + 0] * to_fp32<DataType>(original_input[i + 0]);
```
Comment (ofiryanai) [inline src/VecSim/spaces/computer/preprocessors.h]: P3: Centered L2 no longer consumes x_mean_ip, y_mean_ip, or mean_sum_squares, but WithNorm still computes both dot products and stores an extra float in every L2 storage/query blob. Make correction metadata conditional on Metric == VecSimMetric_IP to avoid dead O(dim) preprocessing work and storage.

File: tests/unit/test_components.cpp [hunk 5]
```diff
@@ -1805,3 +1814,529 @@ INSTANTIATE_TEST_SUITE_P(QuantPreprocessorFP16WithNormTests,
...
+        EXPECT_NEAR(asym_dist, bf, 0.05f) << "Asymmetric L2 vs brute-force";
+
+        allocator->free_allocation(x_blob);
+        allocator->free_allocation(y_blob);
+        allocator->free_allocation(y_query);
+        delete calc;
+    }
+}
+
+TEST(DistanceCalculatorWithNormTest, RandomVectors) {
```
Comment (ofiryanai) [inline tests/unit/test_components.cpp:2279]: P4: The underlying odd-dimension SQ8 kernels are covered by #1001, but the promised odd-dimension DistanceCalculatorWithNorm integration case is still missing. Add one to exercise the complete normalized metadata layout and correction path.

**Review by ofiryanai — DISMISSED** [review]
Blocking correctness and alignment concerns are resolved. The remaining P3/P4 comments are non-blocking.

File: src/VecSim/spaces/computer/preprocessors.h (hunk 4 above)
Comment (dor-forer) [reply src/VecSim/spaces/computer/preprocessors.h]: Fixed in e0a0a203: gated the extra mean_ip metadata slot/computation to `WithNorm && Metric == VecSimMetric_IP` (in `sq8.h` and `preprocessors.h`) instead of `WithNorm` alone. L2 no longer computes or stores `x_mean_ip`/`y_mean_ip` since the centered-L2 correction never reads them. Also had to fix a pre-existing test in `QuantPreprocessorWithNormMetricTest` (1b39f3f5) that was asserting on `mean_ip_index<Metric>()` unconditionally for L2 too — caught by Bugbot after the metadata layout changed size.

File: tests/unit/test_components.cpp (hunk 5 above)
Comment (dor-forer) [reply tests/unit/test_components.cpp:2279]: Added `DistanceCalculatorWithNormTest.OddDimension_MatchesBruteForce` (dim=13) covering IP and L2 symmetric + asymmetric paths through the unaligned storage-blob metadata offset.

**Review by ofiryanai — APPROVED** [review]
